### PR TITLE
[RISCV] Emit VP strided loads/stores in RISCVGatherScatterLowering

### PR DESCRIPTION
**Pull request by lukel97**
RISCVGatherScatterLowering is the last user of riscv_masked_strided_{load,store} after #98131 and #98112, this patch changes it to emit the VP equivalent instead. This allows us to remove the masked_strided intrinsics so we have only have one lowering path.

riscv_masked_strided_{load,store} didn't have AVL operands and were always VLMAX, so this passes in the fixed or scalable element count to the EVL instead, which RISCVVectorPeephole should now convert to VLMAX after #97800.
For loads we also use a vp_select to get passthru (mask undisturbed) behaviour

### CODE REVIEW

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-strided-load-store-asm.ll
Comment (wangpc-pp) [inline llvm/test/CodeGen/RISCV/rvv/fixed-vectors-strided-load-store-asm.ll:147]: Which PR will fix this?

Comment (lukel97) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-strided-load-store-asm.ll:147]: https://github.com/llvm/llvm-project/pull/97798, but I believe @BeMg is working on a CodeGenPrepare version which should handle more cases

Comment (topperc) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-strided-load-store-asm.ll:147]: I messed up who is working on it. It’s @yetingk

Comment (lukel97) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-strided-load-store-asm.ll:147]: In #97798 I had to make sure to use build_vector for fixed-length vectors and splat_vector for scalable vectors.

For CodeGenPrepare does using `IRBuilder::CreateVectorSplat` help? I presume that would get lowered to build_vector and splat_vector for fixed-length and scalable vectors respectively.

Comment (lukel97) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-strided-load-store-asm.ll:147]: Oh I see, but we want the vmv.v.x/vmv.v.f to have vl=evl? Does wrapping the splat in a `@llvm.vp.select` with the passthru and false both set to undef, but using the original mask and evl work? #97989 recently landed which may help set vl=evl

Comment (lukel97) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-strided-load-store-asm.ll:147]: And we can't use `@llvm.riscv.vmv.v.x` because that's only on scalable vectors right?

I think it should still be correct if we splat with AVL=VLMAX because the disabled lanes are poison anyway. It's a bit annoying that there's no good way of controlling the AVL here though but we could leave that to later?

**Comment by preames** [comment]
It looks like all the mentioned changes have landed, is this ready for a rebase?

**Comment by lukel97** [comment]
Rebased and updated to remove `riscv_masked_strided_{load,store}` now that the last user is gone

File: llvm/test/CodeGen/RISCV/rvv/mscatter-combine.ll
Comment (lukel97) [inline llvm/test/CodeGen/RISCV/rvv/mscatter-combine.ll:119]: I think this is because we're missing a generic DAG combine from unit strided VP load/store -> regular load/store. I'm not sure if we have any precedent for canonicalising VP ops to non-VP ops already?

Comment (wangpc-pp) [reply llvm/test/CodeGen/RISCV/rvv/mscatter-combine.ll:119]: Do we need to fix this? The semantics is the same, the difference is vs1r doesn't depend on vl while vse64 does.

Comment (lukel97) [reply llvm/test/CodeGen/RISCV/rvv/mscatter-combine.ll:119]: Yeah, I guess the added dependency on vl is a regression but I'm not overly worried about it, since I don't think the loop vectorizer will emit a unit strided masked gather in the first place. But still happy to address it if others want

Comment (wangpc-pp) [reply llvm/test/CodeGen/RISCV/rvv/mscatter-combine.ll:119]: I'm OK with this, maybe just add a FIXME here? The extra complexity may not be worthy to fix this.

**Review by wangpc-pp — APPROVED** [review]
LGTM.

File: llvm/test/CodeGen/RISCV/rvv/mscatter-combine.ll
Comment (preames) [reply llvm/test/CodeGen/RISCV/rvv/mscatter-combine.ll:121]: Not saying we need to fix this before landing, but we do need to fix this.  There's an existing combine for converting an element strided riscv_maked_strided_* into a masked_*.  We should replicate that over the vp.strided form before calling the transition complete.

Comment (preames) [reply llvm/test/CodeGen/RISCV/rvv/mscatter-combine.ll:121]: (replying to self to continue a thought)

The needed conversion would be from vp.strided.load to a vp.load not a normal or masked load.  Once that was done, we could in turn conversion to a normal load, but that's a separate step.  The first step is simply getting rid of the "strided" part of the intrinsic.

Comment (lukel97) [reply llvm/test/CodeGen/RISCV/rvv/mscatter-combine.ll:121]: Off the top of my head there's already a generic unit strided VP load - > VP load combine, my original comment should have said "full length AVL and all ones mask"

**Comment by lukel97** [comment]
I'll hold off until the LLVM 19 release branch is cut before landing this, I'm not worried about it but it's not particularly urgent either

File: llvm/test/CodeGen/RISCV/rvv/mscatter-combine.ll
Comment (preames) [reply llvm/test/CodeGen/RISCV/rvv/mscatter-combine.ll:121]: Yeah, you're right.  We're getting to the vse64.v, just not the vs1r.  So it's the second step we're missing.

**Review by wangpc-pp — APPROVED** [review]

**Review by preames — APPROVED** [review]
LGTM

**Comment by wangpc-pp** [comment]
Shall we backport these commits?

**Comment by lukel97** [comment]
This patch should be more or less NFC so I don't think it's important to backport it. #100116 might provide some minor optimisations though if people want to see that in LLVM 19

**Comment by lukel97** [comment]
On second thought though this does remove the @llvm.riscv.masked.strided.load and @llvm.riscv.masked.strided.store intrinsics which is breaking. Maybe there's an argument for backporting so we don't have people using them in the meantime.

**Comment by preames** [comment]
I actively oppose backporting this.  Backporting has a risk, and there's no obvious benefit to outweigh that.

**Comment by topperc** [comment]
> On second thought though this does remove the @llvm.riscv.masked.strided.load and @llvm.riscv.masked.strided.store intrinsics which is breaking. Maybe there's an argument for backporting so we don't have people using them between now and LLVM 20

If you believe that somebody external is going to use them or is using them, then we should add AutoUpgrade.cpp support for upgrading them. Independent of whether we backport or not.

**Comment by lukel97** [comment]
I don't think anyone would start using them given that it's not documented and they could just use `@llvm.riscv.vlse.mask`/`@llvm.riscv.vsse.mask` instead. I can't find any existing users at least in the textual format from a quick search on GitHub.

**Pull request merged**
